File: chapter-04-documents/java/DocumentOperations.java
/*
 * DocumentOperations.java
 * Chapter 04: Documents
 *
 * SharePoint Document Operations Example
 * Demonstrates upload, download, and search operations for documents
 */

package com.sharepointexperience.chapter04;

import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.HashMap;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonArray;

/**
 * Example class demonstrating SharePoint document operations
 */
public class DocumentOperations {
    private final String tenantId;
    private final String clientId;
    private final String clientSecret;
    private final HttpClient httpClient;
    private final Gson gson;
    private String accessToken;

    /**
     * Initializes a new instance of DocumentOperations
     */
    public DocumentOperations() {
        this.tenantId = System.getenv("TENANT_ID");
        this.clientId = System.getenv("CLIENT_ID");
        this.clientSecret = System.getenv("CLIENT_SECRET");
        this.httpClient = HttpClient.newHttpClient();
        this.gson = new Gson();

        validateConfig();
    }

    private void validateConfig() {
        if (tenantId == null || clientId == null || clientSecret == null) {
            throw new IllegalArgumentException(
                "TENANT_ID, CLIENT_ID, and CLIENT_SECRET environment variables are required"
            );
        }
    }

    /**
     * Gets access token for Microsoft Graph
     */
    private String getAccessToken() throws IOException, InterruptedException {
        if (accessToken != null) return accessToken;

        String tokenEndpoint = String.format(
            "https://login.microsoftonline.com/%s/oauth2/v2.0/token",
            tenantId
        );

        String form = String.format(
            "client_id=%s&client_secret=%s&scope=%s&grant_type=client_credentials",
            clientId, clientSecret, "https://graph.microsoft.com/.default"
        );

        HttpRequest request = HttpRequest.newBuilder()
            .uri(URI.create(tokenEndpoint))
            .header("Content-Type", "application/x-www-form-urlencoded")
            .POST(HttpRequest.BodyPublishers.ofString(form))
            .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() != 200) {
            throw new IOException("Failed to get access token: " + response.body());
        }

        JsonObject jsonResponse = gson.fromJson(response.body(), JsonObject.class);
        accessToken = jsonResponse.get("access_token").getAsString();
        return accessToken;
    }

    /**
     * Uploads a file to SharePoint
     */
    public void uploadFile(String siteId, String driveId, String filePath, String destinationFileName)
            throws IOException, InterruptedException {
        try {
            System.out.println("Uploading file: " + filePath);

            Path path = Paths.get(filePath);
            if (!Files.exists(path)) {
                throw new IOException("File not found: " + filePath);
            }

            byte[] fileContent = Files.readAllBytes(path);
            String encodedName = java.net.URLEncoder.encode(destinationFileName, "UTF-8");

            String token = getAccessToken();
            String url = String.format(
                "https://graph.microsoft.com/v1.0/sites/%s/drives/%s/items/root:/%s:/content",
                siteId, driveId, encodedName
            );

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/octet-stream")
                .PUT(HttpRequest.BodyPublishers.ofByteArray(fileContent))
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200 && response.statusCode() != 201) {
                throw new IOException("Failed to upload file: " + response.body());
            }

            JsonObject result = gson.fromJson(response.body(), JsonObject.class);
            System.out.println("File uploaded successfully:");
            System.out.println("  ID: " + result.get("id").getAsString());
            System.out.println("  Name: " + result.get("name").getAsString());
            System.out.println("  Size: " + result.get("size").getAsLong() + " bytes");
            System.out.println("  Web URL: " + result.get("webUrl").getAsString());

        } catch (Exception e) {
            System.err.println("Error uploading file: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Downloads a file from SharePoint
     */
    public void downloadFile(String siteId, String driveId, String itemId, String downloadPath)
            throws IOException, InterruptedException {
        try {
            System.out.println("Downloading file to: " + downloadPath);

            String token = getAccessToken();
            String url = String.format(
                "https://graph.microsoft.com/v1.0/sites/%s/drives/%s/items/%s/content",
                siteId, driveId, itemId
            );

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .GET()
                .build();

            HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());

            if (response.statusCode() != 200) {
                throw new IOException("Failed to download file: " + response.statusCode());
            }

            try (FileOutputStream fos = new FileOutputStream(downloadPath)) {
                fos.write(response.body());
            }

            System.out.println("File downloaded successfully to: " + downloadPath);

        } catch (Exception e) {
            System.err.println("Error downloading file: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Lists files in a specific folder
     */
    public void listFilesInFolder(String siteId, String driveId, String folderPath)
            throws IOException, InterruptedException {
        try {
            System.out.println("Listing files in folder: " + (folderPath.isEmpty() ? "root" : folderPath));

            String token = getAccessToken();
            String url;

            if (folderPath.isEmpty()) {
                url = String.format(
                    "https://graph.microsoft.com/v1.0/sites/%s/drives/%s/root/children",
                    siteId, driveId
                );
            } else {
                String encodedPath = java.net.URLEncoder.encode(folderPath, "UTF-8");
                url = String.format(
                    "https://graph.microsoft.com/v1.0/sites/%s/drives/%s/items/root:/%s:/children",
                    siteId, driveId, encodedPath
                );
            }

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() != 200) {
                throw new IOException("Failed to list files: " + response.body());
            }

            JsonObject jsonResponse = gson.fromJson(response.body(), JsonObject.class);
            JsonArray items = jsonResponse.getAsJsonArray("value");

            System.out.println("Found " + items.size() + " items:");
            System.out.println("-".repeat(80));

            for (int i = 0; i < items.size(); i++) {
                JsonObject item = items.get(i).getAsJsonObject();
                String itemType = item.has("folder") ? "Folder" : "File";
                System.out.println(itemType + ": " + item.get("name").getAsString());
                System.out.println("  ID: " + item.get("id").getAsString());
                System.out.println("  Size: " + item.get("size").getAsLong() + " bytes");
                System.out.println("-".repeat(80));
            }

        } catch (Exception e) {
            System.err.println("Error listing files: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Lists the document libraries (drives) of a site. Solo lectura.
     */
    public void listDrives(String siteId) throws IOException, InterruptedException {
        try {
            System.out.println("Listing document libraries of site: " + siteId);
            String token = getAccessToken();
            String url = String.format("https://graph.microsoft.com/v1.0/sites/%s/drives", siteId);

            HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .header("Content-Type", "application/json")
                .GET()
                .build();

            HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() != 200) {
                throw new IOException("Error listing drives: " + response.body());
            }

            JsonObject json = gson.fromJson(response.body(), JsonObject.class);
            var drives = gson.fromJson(json.get("value"), com.google.gson.JsonArray.class);
            System.out.println("Found " + drives.size() + " libraries:");
            System.out.println("-".repeat(80));
            for (var d : drives) {
                JsonObject drive = d.getAsJsonObject();
                System.out.println("Library: " + drive.get("name"));
                System.out.println("  ID: " + drive.get("id"));
                System.out.println("  Type: " + drive.get("driveType"));
                System.out.println("-".repeat(80));
            }
        } catch (Exception e) {
            System.err.println("Error listing libraries: " + e.getMessage());
            throw e;
        }
    }

    /**
     * Entry point: demo de solo lectura contra el sitio de pruebas book-test.
     */
    public static void main(String[] args) {
        try {
            System.out.println("=== SharePoint Document Operations Example ===\n");
            DocumentOperations docOps = new DocumentOperations();

            String hostname = System.getenv("SHAREPOINT_HOSTNAME");
            if (hostname == null || hostname.isBlank()) hostname = "olddogsoft1.sharepoint.com";
            String sitePath = System.getenv("SHAREPOINT_SITE_PATH");
            if (sitePath == null || sitePath.isBlank()) sitePath = "book-test";

            // Resolver el sitio por path para obtener su ID.
            String token = docOps.getAccessToken();
            String siteUrl = String.format(
                "https://graph.microsoft.com/v1.0/sites/%s:/sites/%s",
                java.net.URLEncoder.encode(hostname, "UTF-8"),
                java.net.URLEncoder.encode(sitePath, "UTF-8"));
            HttpRequest siteReq = HttpRequest.newBuilder()
                .uri(URI.create(siteUrl))
                .header("Authorization", "Bearer " + token)
                .GET()
                .build();
            HttpResponse<String> siteResp = docOps.httpClient.send(siteReq, HttpResponse.BodyHandlers.ofString());
            if (siteResp.statusCode() != 200) {
                throw new IOException("Error resolving site: " + siteResp.body());
            }
            String siteId = docOps.gson.fromJson(siteResp.body(), JsonObject.class).get("id").getAsString();

            docOps.listDrives(siteId);
            System.out.println("\nDocument operations completed successfully!");

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: chapter-04-documents/csharp/DocumentOperations.cs
/*
 * DocumentOperations.cs
 * Chapter 04: Documents
 *
 * SharePoint Document Operations Example
 * Demonstrates upload, download, and search operations for documents
 */

using System;
using System.IO;
using System.Threading.Tasks;
using LaExperiencia.SharePoint.Common;
using Microsoft.Graph;
using Microsoft.Graph.Models;

namespace LaExperiencia.SharePoint.Chapter04.Documents
{
    /// <summary>
    /// Example class demonstrating SharePoint document operations.
    /// The Graph client is injected via constructor (DI) from the common module.
    /// </summary>
    public class DocumentOperations
    {
        private readonly GraphServiceClient _graphClient;

        /// <summary>Creates an instance with an injected Graph client (DI).</summary>
        public DocumentOperations(GraphServiceClient graphClient)
        {
            _graphClient = graphClient ?? throw new ArgumentNullException(nameof(graphClient));
        }

        /// <summary>
        /// Uploads a file to a SharePoint document library
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="filePath">Local path to the file</param>
        /// <param name="destinationFileName">Name for the file in SharePoint</param>
        /// <returns>The uploaded drive item</returns>
        public async Task<DriveItem?> UploadFileAsync(
            string siteId,
            string driveId,
            string filePath,
            string destinationFileName)
        {
            try
            {
                Console.WriteLine($"Uploading file: {filePath}");

                // Read file content
                byte[] fileContent = await File.ReadAllBytesAsync(filePath);
                using var stream = new MemoryStream(fileContent);

                // Upload the file
                var uploadedItem = await _graphClient
                    .Drives[driveId]
                    .Items["root"]
                    .ItemWithPath(destinationFileName)
                    .Content
                    .PutAsync(stream);

                Console.WriteLine($"File uploaded successfully:");
                Console.WriteLine($"  ID: {uploadedItem?.Id}");
                Console.WriteLine($"  Name: {uploadedItem?.Name}");
                Console.WriteLine($"  Size: {uploadedItem?.Size} bytes");
                Console.WriteLine($"  Web URL: {uploadedItem?.WebUrl}");

                return uploadedItem;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error uploading file: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Downloads a file from SharePoint
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID of the file</param>
        /// <param name="downloadPath">Local path to save the file</param>
        public async Task DownloadFileAsync(
            string siteId,
            string driveId,
            string itemId,
            string downloadPath)
        {
            try
            {
                Console.WriteLine($"Downloading file to: {downloadPath}");

                // Get the file content
                var stream = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .Content
                    .GetAsync();

                if (stream == null)
                {
                    throw new InvalidOperationException("Graph returned no content for the requested item.");
                }

                // Save to local file
                using var fileStream = new FileStream(downloadPath, FileMode.Create);
                await stream.CopyToAsync(fileStream);

                Console.WriteLine($"File downloaded successfully to: {downloadPath}");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error downloading file: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Searches for files across SharePoint
        /// </summary>
        /// <param name="query">Search query</param>
        public async Task SearchFilesAsync(string query)
        {
            try
            {
                Console.WriteLine($"Searching for files: '{query}'");

                // Search for drive items
                var searchResults = await _graphClient.Search.Query.PostAsQueryPostResponseAsync(new Microsoft.Graph.Search.Query.QueryPostRequestBody
                {
                    Requests = new System.Collections.Generic.List<Microsoft.Graph.Models.SearchRequest>
                    {
                        new Microsoft.Graph.Models.SearchRequest
                        {
                            EntityTypes = new System.Collections.Generic.List<Microsoft.Graph.Models.EntityType?>
                            {
                                Microsoft.Graph.Models.EntityType.DriveItem
                            },
                            Query = new Microsoft.Graph.Models.SearchQuery
                            {
                                QueryString = query
                            }
                        }
                    }
                });

                Console.WriteLine($"Search results:");
                Console.WriteLine(new string('-', 80));

                foreach (var result in searchResults?.Value ?? new System.Collections.Generic.List<Microsoft.Graph.Models.SearchResponse>())
                {
                    foreach (var hit in result?.HitsContainers ?? new System.Collections.Generic.List<Microsoft.Graph.Models.SearchHitsContainer>())
                    {
                        foreach (var item in hit?.Hits ?? new System.Collections.Generic.List<Microsoft.Graph.Models.SearchHit>())
                        {
                            var resource = item?.Resource as DriveItem;
                            if (resource != null)
                            {
                                Console.WriteLine($"Name: {resource.Name}");
                                Console.WriteLine($"  Web URL: {resource.WebUrl}");
                                Console.WriteLine($"  Size: {resource.Size} bytes");
                                Console.WriteLine(new string('-', 80));
                            }
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error searching files: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Lists files in a specific folder
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="folderPath">Path to the folder</param>
        public async Task ListFilesInFolderAsync(string siteId, string driveId, string folderPath)
        {
            try
            {
                Console.WriteLine($"Listing files in folder: {folderPath}");

                var items = await _graphClient
                    .Drives[driveId]
                    .Root
                    .ItemWithPath(folderPath)
                    .Children
                    .GetAsync();

                Console.WriteLine($"Found {items?.Value?.Count ?? 0} items:");
                Console.WriteLine(new string('-', 80));

                foreach (var item in items?.Value ?? new System.Collections.Generic.List<DriveItem>())
                {
                    string itemType = item.Folder != null ? "Folder" : "File";
                    Console.WriteLine($"{itemType}: {item.Name}");
                    Console.WriteLine($"  ID: {item.Id}");
                    Console.WriteLine($"  Size: {item.Size} bytes");
                    Console.WriteLine($"  Last Modified: {item.LastModifiedDateTime}");
                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error listing files: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Gets file metadata
        /// </summary>
        /// <param name="siteId">The site ID</param>
        /// <param name="driveId">The drive ID</param>
        /// <param name="itemId">The item ID</param>
        /// <returns>The drive item metadata</returns>
        public async Task<DriveItem?> GetFileMetadataAsync(string siteId, string driveId, string itemId)
        {
            try
            {
                Console.WriteLine($"Fetching metadata for item: {itemId}");

                var item = await _graphClient
                    .Drives[driveId]
                    .Items[itemId]
                    .GetAsync();

                Console.WriteLine($"File metadata:");
                Console.WriteLine($"  Name: {item?.Name}");
                Console.WriteLine($"  ID: {item?.Id}");
                Console.WriteLine($"  Size: {item?.Size} bytes");
                Console.WriteLine($"  Created: {item?.CreatedDateTime}");
                Console.WriteLine($"  Modified: {item?.LastModifiedDateTime}");
                Console.WriteLine($"  Web URL: {item?.WebUrl}");

                return item;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error getting file metadata: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Entry point for the example
        /// <summary>
        /// Lists the document libraries (drives) of a site. Read-only.
        /// </summary>
        public async Task ListDrivesAsync(string siteId)
        {
            try
            {
                Console.WriteLine($"Listing document libraries of site: {siteId}");
                var drives = await _graphClient.Sites[siteId].Drives.GetAsync();

                Console.WriteLine($"Found {drives?.Value?.Count ?? 0} libraries:");
                Console.WriteLine(new string('-', 80));
                foreach (var drive in drives?.Value ?? new System.Collections.Generic.List<Drive>())
                {
                    Console.WriteLine($"Library: {drive.Name}");
                    Console.WriteLine($"  ID: {drive.Id}");
                    Console.WriteLine($"  Type: {drive.DriveType}");
                    Console.WriteLine(new string('-', 80));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error listing libraries: {ex.Message}");
                throw;
            }
        }

        /// <summary>
        /// Entry point: builds the client with the common module and runs a read-only
        /// demo listing the libraries of the book-test test site.
        /// </summary>
        public static async Task Main(string[] args)
        {
            try
            {
                var graphClient = SharePointGraphClientFactory.CreateFromSecret();
                var docOps = new DocumentOperations(graphClient);

                var hostname = Environment.GetEnvironmentVariable("SHAREPOINT_HOSTNAME") ?? "olddogsoft1.sharepoint.com";
                var sitePath = Environment.GetEnvironmentVariable("SHAREPOINT_SITE_PATH") ?? "book-test";

                // Resolve the site by path to get its ID (the path form does not work for .Drives).
                var site = await graphClient.Sites[$"{hostname}:/sites/{sitePath}"].GetAsync();
                if (site == null)
                {
                    Console.WriteLine("Test site not found.");
                    return;
                }
                var siteId = site.Id ?? throw new InvalidOperationException("El sitio no tiene ID.");
                await docOps.ListDrivesAsync(siteId);

                Console.WriteLine("\nDocument operations completed successfully!");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
                Environment.Exit(1);
            }
        }
    }
}
